File: hackerrank/problem solving/algorithms/2 - implementation/24 - viral advertising/viral_advertising_oo.cpp
// Source: https://www.hackerrank.com/challenges/strange-advertising/problem?isFullScreen=true

#include <iostream>

using namespace std;


class Viral_Advertising {

private:
    int _days;
    int _cumulative_likes;

    void _viral_advertising() {
        int shared {5}, liked {0};

        for (int _ {0}; _ < _days; _++) {
            liked = {shared / 2};
            _cumulative_likes += liked;
            shared = {liked * 3};
        }
    }

public:
    Viral_Advertising(int days): _days{days} {
        _cumulative_likes = {0};
        _viral_advertising();
        print_cumulative_likes();
    }

        void print_cumulative_likes() const {
            cout << _cumulative_likes << "\n";
        }
};

int main() {
    int days;
    cin >> days;
    const Viral_Advertising cumulative_likes(days);

    return 0;
}

File: hackerrank/problem solving/algorithms/2 - implementation/24 - viral advertising/viral_advertising_oo.js
// Source: https://www.hackerrank.com/challenges/strange-advertising/problem?isFullScreen=true

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');
    main();
});

function readLine() {
    return inputString[currentLine++];
}


function main() {
    const days = +readLine();
    const cumulativeLikes = new ViralAdvertising(days);
}

    class ViralAdvertising {
        #days;
        #cumulativeLikes;

        constructor(days) {
            this.#days = days;
            this.#cumulativeLikes = 0;
        
            this.#viralAdvertising();
            this.printCumulativeLikes();
        }

        #viralAdvertising() {
            let [shared, liked, cumulative] = [5, 0, 0];

            for (let _ = 0; _ < this.#days; _++) {
                liked = Math.trunc(shared / 2);
                this.#cumulativeLikes += liked;
                shared = liked * 3;
            }
        }

        printCumulativeLikes() {
            console.log(this.#cumulativeLikes);
        }
    }
